working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+prefixed by `protected`. We say that it is a _protected extension type
+declaration_, and that it introduces a _protected extension type_.
+
+Let `E` be the name of a protected extension type declaration.
+The type `E` (or
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>
+if `E` is generic) is a proper subtype of `Object?` and a proper supertype
+of `Never`.
+
+*In contrast to non-protected extension types, the extension type has no
+subtype relationship with its on-type.*
+
+It is a compile-time error if a protected extension type `E` is the
+target type in a type test (`o is E` respectively
+<code>o is E<T<sub>1</sub>, .. T<sub>k</sub>></code>)
+or a type cast (`o as E` respectively
+<code>o as E<T<sub>1</sub>, .. T<sub>k</sub>></code>).
+The type `dynamic` is not assignable to any protected extension type.
+
+*The subtype relationships and the type test/cast errors ensure that an
+instance creation expression is the only way to create values of a
+protected extension type `E`: There is no assignability from the on-type to
+`E`, and it is an error to cast or promote an expression to `E`.*
+
+*This is a crucial property, because it ensures that the value of an
+expression with static type `E` has been obtained as the return value of an
+extension type constructor, and that allows us to write arbitrary code that
+ensures that an object typed as `E` satisfies certain constraints. For
+instance, if all constructors of `E` ensure that a given invariant holds,
+and if every member of `E` preserves that invariant, and if there are no
+aliases to the underlying instance of the on-type of `E` typed as any other
+type than `E`, then the invariant is guaranteed to be preserved.*
+
+It is a compile-time error for an extension declaration to start with the
+keyword `protected`, unless it is explicit (*that is, unless it also has
+the keyword `type`*).
+
+*A protected extension type should not allow for implicit extension method
+invocations, because they are inherently not guarded by the execution of an
+extension type constructor.*
+
+*For example:*
+
+```dart
+protected extension type nat on int {
+  factory nat(int value) =>
+      value >= 0 ? value : throw "Attempt to create an invalid nat";
+}
+
+void main() {
+  nat n1 = 42; // Error.
+  var n2 = nat(42); // OK at compile time, and at run time.
+  var n3 = nat(-1); // OK at compile time, throws at run time.
+}
+```
+
+*The following example illustrates the subtyping relationships; it
+illustrates the use of a constructor to enforce an invariant (that the
+`int` in the `IntBox` is even); and it illustrates that the available
+methods (there is just one: `next()`) all preserve that invariant.*
+
+```dart
+class IntBox {
+  int i;
+  IntBox(this.i);
+}
+
+protected extension type EvenIntBox on IntBox {
+  factory EvenIntBox(int i) =>
+      i.isEven ? IntBox(i) : throw "Invalid EvenIntBox";
+  factory EvenIntBox.fromIntBox(IntBox intBox) =>
```

Comment:
I feel like this example would be stronger without this factory, since it means that you're not enforcing the invariant here at all.

Reply:
True, but I think it's only fair to mention that the invariant preservation relies on manual reasoning by developers: If there is an alias to a mutable object then a reference like `intBox` in `main` can be used to violate the invariant.

The point is that the type incompatibility (statically for the on-type directly, and both statically and dynamically whenever the type is used as a type argument) helps developers centralize the code that maintains the invariant to a well-known, small amount of code. There are lots of loopholes, but a developer who wants to avoid them will have an easier type doing it than they would if we had direct assignability (like any non-protected extension type with no constructors).